tools/vswhere/MSBuildLocator.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Xamarin.Android.Tools.VSWhere
+{
+	/// <summary>
+	/// https://github.com/Microsoft/vswhere/wiki/Find-MSBuild
+	/// </summary>
+	public class MSBuildLocator
+	{
+		static readonly string [] msbuildLocations = new [] {
+			// VS 2019 & Higher
+			Path.Combine ("MSBuild", "Current", "Bin", "MSBuild.exe"),
+			// VS 2017
+			Path.Combine ("MSBuild", "15.0", "Bin", "MSBuild.exe"),
+		};
+
+		/// <summary>
+		/// Adds the -prerelease flag, defaults to false
+		/// </summary>
+		public bool IncludePrerelease { get; set; } = false;
+
+		public string VisualStudioDirectory { get; private set; }
+
+		public string MSBuildPath { get; private set; }
+
+		public void Locate ()
+		{
+			var vsInstallDir = Environment.GetEnvironmentVariable ("VSINSTALLDIR");
+			if (string.IsNullOrEmpty (vsInstallDir)) {
+				var programFiles = Environment.GetFolderPath (Environment.SpecialFolder.ProgramFilesX86);
+				var vswhere = Path.Combine (programFiles, "Microsoft Visual Studio", "Installer", "vswhere.exe");
+				if (!File.Exists (vswhere))
+					throw new FileNotFoundException ("Cannot find vswhere.exe!", vswhere);
+
+				string prerelease = IncludePrerelease ? "-prerelease" : "";
+				VisualStudioDirectory = Exec (vswhere, $"-latest {prerelease} -products * -requires Microsoft.Component.MSBuild -property installationPath");
+				if (!Directory.Exists (VisualStudioDirectory)) {
+					throw new Exception ($"vswhere.exe result returned a directory that did not exist: {VisualStudioDirectory}");
+				}
+			} else {
+				VisualStudioDirectory = vsInstallDir;
+			}
+
+			foreach (var path in msbuildLocations) {
+				MSBuildPath = Path.Combine (VisualStudioDirectory, path);
+				if (File.Exists (MSBuildPath))
+					return;
+			}
+
+			throw new FileNotFoundException ("Cannot find MSBuild.exe!");
+		}
+
+		string Exec (string fileName, string args)
+		{
+			var info = new ProcessStartInfo {
```

Comment:
@grendello? Is this `ProcessStartInfo` usage safe?

Reply:
@grendello: related: could you write up a doc to describe how to *safely* use `ProcessStartInfo`, and the things to look out for?

Reply:
FYI @grendello the command does this:
```
>"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -latest -requires Microsoft.Component.MSBuild -property installationPath
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise
```
It prints a one-line path to stdout.

Reply:
Yes, this kind of usage is safe. You might however want to use a timeout when waiting for the process - it's always safer this way.